tests/kernel/mem_protect/mpu/src/kobject.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_mpu.h"
+
+/* Kernel objects */
+K_THREAD_STACK_DEFINE(kobject_test_4_stack, KOBJECT_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kobject_test_5_1_stack, KOBJECT_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kobject_test_5_2_stack, KOBJECT_STACK_SIZE);
+
+K_SEM_DEFINE(kobject_sem, SEMAPHORE_INIT_COUNT, SEMAPHORE_MAX_COUNT);
+K_MUTEX_DEFINE(kobject_mutex);
+__kernel struct k_thread kobject_test_4_tid;
+__kernel struct k_thread kobject_test_5_1_tid, kobject_test_5_2_tid;
+struct k_sem *random_sem_type=NULL;
+
+
+/****************************************************************************/
+void kobject_user_tc1(void *p1, void *p2, void *p3)
+{
+	valid_fault= true;
+	k_sem_take(random_sem_type, K_FOREVER);
+
+
+}
+
+/* Test access to a invalid semaphore who's address is NULL. */
+void kobject_test_1(void *p1, void *p2, void *p3)
+{
+
+	/* _k_object_init(random_sem_type); */
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex,
+			      random_sem_type, NULL);
+
+	k_thread_user_mode_enter(kobject_user_tc1, NULL, NULL, NULL);
+
+}
+/****************************************************************************/
+void kobject_user_tc2(void *p1, void *p2, void *p3)
+{
+	valid_fault= false;
+	k_sem_give(&kobject_sem);
+
+	/* should cause a fault */
+	valid_fault= true;
+	/* typecasting to override compiler warning */
+	k_sem_take((struct k_sem *)&kobject_mutex, K_FOREVER);
+
+
+}
+
+/* Test if a syscall can take a different type of kobject. */
+void kobject_test_2(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex, NULL);
+
+	k_thread_user_mode_enter(kobject_user_tc2, NULL, NULL, NULL);
+
+}
+
+/****************************************************************************/
+void kobject_user_tc3(void *p1, void *p2, void *p3)
```

Comment:
can we give more descriptive names to these cases than stuff like "tc3"?

Reply:
This would be really hard for me to do. I have many functions in this file. To give meaningful unique names to each of them is hard.